src/stan_math_backend/Statement_gen.ml

```diff
@@ -75,9 +69,35 @@ let trans_math_fn fname =
     value ~default:(fname, [])
       (bind (internal_fn_of_string fname) ~f:math_fn_translations))
 
+let rec block_wrapping {stmt; smeta} =
+  let in_block {stmt; smeta} =
+    { stmt=
+        ( match stmt with
+        | Block l | SList l -> Block l
+        | stmt -> Block [{stmt; smeta}] )
+    ; smeta }
+  in
+  let block_wrapping_base stmt =
+    match stmt with
+    | IfElse (_, _, _) | While (_, _) | For _ ->
+        map_statement (fun x -> x) in_block stmt
+    | _ -> stmt
+  in
+  { stmt= block_wrapping_base (map_statement (fun x -> x) block_wrapping stmt)
+  ; smeta }
+
 let rec pp_statement (ppf : Format.formatter)
     ({stmt; smeta} : (mtype_loc_ad, 'a) stmt_with) =
+  let {stmt; smeta} = block_wrapping {stmt; smeta} in
+  (* Make sure that all statements are safely wrapped in a block in such a way that we can insert a location update before.
+     Note that this should not change the semantics as we are only inserting blocks around the whole body of a for-, while- or if-body. 
```

Comment:
Excellent comment! We can merge this but I might suggest I rename the function to something like `block_wrap_stmt_bodies` or maybe `ensure_body_bock` - I think I failed to notice on first read through that it was just adding them for those 3 statement types and only if they were missing, which is key. My bad.

Reply:
Absolutely. Go ahead and rename it!